dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java

```diff
@@ -0,0 +1,270 @@
+        for (String s : list) {
+            if (s.indexOf('/') > 0) {
+                Tr.warning(tc, INVALID_APP_WARNING, s, elementName);
+            }
+        }
+    }
+
+    private static class MpConfigServerConfigObject implements OpenAPIServerXMLConfig {
+
+        private final List<String> includedApps = new ArrayList<String>();
+        private final List<String> excludedApps = new ArrayList<String>();
+
+        private final List<String> includedModules = new ArrayList<String>();
+        private final List<String> excludedModules = new ArrayList<String>();
+
+        private Optional<ConfigMode> configMode = Optional.empty();
+
+        private MpConfigServerConfigObject(Map<String, Object> properties) {
+            //Read the server.xml
+            Optional<String[]> includedAppArray = Optional.ofNullable((String[]) properties.get(INCLUDE_APP_PROPERTY_NAME));
+            Optional<String[]> excludedAppArray = Optional.ofNullable((String[]) properties.get(EXCLUDE_APP_PROPERTY_NAME));
+            Optional<String[]> includedModulesArray = Optional.ofNullable((String[]) properties.get(INCLUDE_MODULE_PROPERTY_NAME));
+            Optional<String[]> excludedModulesArray = Optional.ofNullable((String[]) properties.get(EXCLUDE_MODULE_PROPERTY_NAME));
+
+            Optional<List<String>> includedAppList = includedAppArray.map(Arrays::asList);
+            Optional<List<String>> excludedAppList = excludedAppArray.map(Arrays::asList);
+            Optional<List<String>> includedModulesList = includedModulesArray.map(Arrays::asList);
+            Optional<List<String>> excludedModulesList = excludedModulesArray.map(Arrays::asList);
+
+            //Send a warning message if the contents are malformed
+            includedAppList.ifPresent((List<String> list) -> appStringProcessing(list, INCLUDE_APP_PROPERTY_NAME));
+            excludedAppList.ifPresent((List<String> list) -> appStringProcessing(list, EXCLUDE_APP_PROPERTY_NAME));
+            includedModulesList.ifPresent((List<String> list) -> warnInvalidModules(list, INCLUDE_MODULE_PROPERTY_NAME));
+            excludedModulesList.ifPresent((List<String> list) -> warnInvalidModules(list, EXCLUDE_MODULE_PROPERTY_NAME));
+
+            if (includedAppList.isPresent() && includedAppList.get().size() == 1) {
+                String includedString = includedAppList.get().get(0).toLowerCase().trim();
+                if (includedString.equals("all")) {
+                    configMode = Optional.of(ConfigMode.All);
+                    includedAppList = Optional.empty();
+                } else if (includedString.equals("first")) {
+                    configMode = Optional.of(ConfigMode.First);
+                    includedAppList = Optional.empty();
+                } else if (includedString.equals("none")) {
+                    configMode = Optional.of(ConfigMode.None);
+                    includedAppList = Optional.empty();
+                }
+            }
```

Comment:
This logic seems questionable. Consider:

```xml
<mpOpenAPI>
  <includeApplication>all</includeApplication>
  <includeApplication>test1</includeApplication>
</mpOpenAPI>
```

With the current code, I think only `test1` would be included, and we'd treat `all` as an application name and emit a confusing warning.

Could be a follow up issue, but it needs resolved before release.

Reply:
Discussed in slack, we decided we'll need to think further on what's correct here.

Reply:
Still uncertain about this. I don't like that the meaning of `<includeApplication>all</includeApplication>` changes when you add another `<includeApplication>` element.

Will take another look in a follow-up issue.